Automattic/jetpack
# Blocks: Add a new shared Button component

See #14935

Proposal for a new shared `Button` with:
- `ButtonEdit` component to be used in the `edit` part of a block.
- `ButtonSave` component to be used in the `save` part of a block.
- `getButtonEditProps` utility to only pass the relevant block's props to `ButtonEdit`.
- `getButtonAttributes` utility to add all the `Button` attributes to the block's attributes definition.
- `jetpack_get_button_classes` utility to get all the `Button` classes for SSR.
- `jetpack_get_button_styles` utility to get all the `Button` styles for SSR.

### The Background

Many Jetpack blocks use buttons and would love to have all the Core `Button` block niceties (colors, border radius, gradients, styles, etc.), but there's no straightforward way to reuse the Core `Button`.
- Core `Button` does not export `ButtonEdit` and `ButtonSave`.
- Core `Button` is in fact a link that looks like a button, and have an UI to set the `url` attribute.
- Core `Button` is a standalone block, so its attributes refer to itself (e.g. `backgroundColor` can only be the button's background color), while in our blocks the `Button` is often just part of a block (is `backgroundColor` the button's background color, or the whole block's?).
- Jetpack blocks use Buttons in many different ways: some server-side render it; some save it in the post content; some need an ID or extra attributes; they can be `a`, `button`, `input` elements; they can have an `onClick` handler; etc.

There are many possible solutions for each of this issues, but as far as I can see (and I might very well be wrong!) the Core `Button` would not fill _all_ our needs, and even if it did, the Jetpack minimum compatibility would mean not being able to use new Core changes for months.

For these reasons, we tried re-implementing the Core `Button` block in-house, and then we did it again, and again.
As far as I can see, we have (at least?) 3 different implementation of a button component, and no actual standard.

The `Button` component proposed is meant to be a catch-all solution, as close as possible to Core's `Button` block, but also extendable for any of our niche needs and easy enough to use as a drop-in replacement of our current implementations.

### The Ideas

1. Start from today's Core `Button`, so to have all of its most recent features.

2. Instead of straight-up copy-pasting from Core, restructure the code in smaller files, that should be more manageable and easier to keep up to date with Core changes.

3. Make the attribute naming consistent: all `Button` attributes must start with `button` (e.g. `buttonBackgroundColor`), with the sole exception of custom color attributes, that are supposed to be prepended with `custom` (e.g. `customButtonBackgroundColor`).

4. Create a handful of utilities that would drastically decrease code duplication, and promote consistency.

### Other Examples

#14931

---

#### Changes proposed in this Pull Request:

* Create a new `Button` component, complete with helpers for all common use cases, to replace the several different re-implementation of Core's `Button` block.
* Use the new `Button` component in the `Revue` block.

#### Testing instructions:






* Before applying this PR, create a new post with several Revue blocks.
  * One with named colors and gradients
  * One with custom colors.
  * One with modified button text and labels.
* Save the post!
* Apply this PR and reload the post.
* Make sure all the Revue blocks look as expected in both editor and front end.
* Add another Revue block from scratch, and try all the various button settings.
* Make sure they work as expected in both editor and front end.

#### Proposed changelog entry for your changes:

N/A
